docs/topics/typecasts.md

```diff
@@ -252,30 +408,176 @@ Smart casts can be used in the following conditions:
     </tr>
 </table>
 
-## "Unsafe" cast operator
+### Cast operators {id="unsafe-cast-operator"}  {id="safe-nullable-cast-operator"}
 
-To explicitly cast an object to a non-nullable type, use the *unsafe* cast operator `as`:
+Kotlin has two cast operators: `as` and `as?`. You can use both to cast, but they have different behaviors.
+
+If a cast fails with the `as` operator, a `ClassCastException` is thrown at runtime. That's why it's also called the **unsafe** operator.
+You can use `as` when casting to a non-null type:
 
 ```kotlin
-val x: String = y as String
+fun main() {
+    val rawInput: Any = "user-1234"
+
+    // Casts to String successfully
+    val userId = rawInput as String
+    println("Logging in user with ID: $userId")
+    // Logging in user with ID: user-1234
+
+    // Triggers ClassCastException
+    val wrongCast = rawInput as Int
+    println("wrongCast contains: $wrongCast")
+    // Exception in thread "main" java.lang.ClassCastException
+}
 ```
+{kotlin-runnable="true" kotlin-min-compiler-version="2.0" id="kotlin-unsafe-cast-operator" validate="false"}
 
-If the cast isn't possible, the compiler throws an exception. This is why it's called _unsafe_.
+If you use the `as?` operator instead, and the cast fails, the operator returns `null`. That's why it's also
+called the **safe** operator:
 
-In the previous example, if `y` is `null`, the code above also throws an exception. This is because `null` can't be cast
-to `String`, as `String` isn't [nullable](null-safety.md). To make the example work for possible null values, use a nullable
-type on the right-hand side of the cast:
+```kotlin
+fun main() {
+    val rawInput: Any = "user-1234"
+
+    // Casts to String successfully
+    val userId = rawInput as? String
+    println("Logging in user with ID: $userId")
+    // Logging in user with ID: user-1234
+
+    // Assigns a null value to wrongCast
+    val wrongCast = rawInput as? Int
+    println("wrongCast contains: $wrongCast")
+    // wrongCast contains: null
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="2.0" id="kotlin-safe-cast-operator"}
+
+#### Nullable types
```

Comment:
I think for consistency, `as` and `as?` operators should either have each a corresponding subsection, or none at all. Nullable types is rather short and could be just a part of Casting operators. But, again, if we want them to be visible in the right menu, two subheaders would be better.

Reply:
Not sure what you mean by consistency? Do you mean so it's similar in naming to  the "`is` and `!is` operators" header? Previously we had the `as` and `as?` operators described in two separate sections but a common complaint was that this separation made it harder to understand what the difference in behavior is. Users wanted to see one example showing the behavior of both. So that's why I put them in one section.

Nullable types has its own section to try and help users find it in TOC and search. It's important for Java developers converting code to Kotlin because the least effort approach is to use `as` but we want them to use `as?`.

Reply:
OK. I removed the "nullable types" header but kept the"Up and downcasting" section. Is that better?

Reply:
Thank you! Another suggestion, maybe mention the `as` and `as?` operators in the "Cast operators" header? like in "Checks with `is` and `!is` operators"